src/frontend/Semantic_check.ml

```diff
@@ -314,6 +316,47 @@ let semantic_check_fn_stan_math ~is_cond_dist ~loc id es =
       |> Semantic_error.illtyped_stanlib_fn_app loc id.name
       |> Validate.error
 
+let semantic_check_reduce_sum ~is_cond_dist ~loc id es =
+  let arg_match (x_ad, x_t) y =
+    UnsizedType.check_of_same_type_mod_conv "" x_t y.emeta.type_
+    && UnsizedType.autodifftype_can_convert x_ad y.emeta.ad_level
+  in
+  let args_match a b =
+    if List.length a = List.length b then List.for_all2_exn ~f:arg_match a b
+    else false
+  in
+  let return_generic_error =
+    es
+    |> List.map ~f:type_of_expr_typed
+    |> Semantic_error.illtyped_reduce_sum_generic loc id.name
+    |> Validate.error
+  in
+  match es with
+  | { emeta=
+        { type_=
+            UnsizedType.UFun
+              ( (_, UInt)
+                :: (_, UInt)
+                   :: ((_, sliced_arg_fun_type) as sliced_arg_fun) :: fun_args
+              , ReturnType UReal ); _ }; _ }
```

Comment:
I wonder if the concrete types here are too strict - for example, do we want to allow functions that return Int, since they can be converted? In any case, too strict is probably fine.

Reply:
Yeah, here I was debating with myself whether to allow int. I dont have an opinion. @wds15 or @bbbales2 do you guys have one?

Reply:
Hmm...never thought about this. Let’s check if we tested that in math and decide based on that? I really need to look that up.

Reply:
I think only real for the function return type.

Reply:
We'd never autodiff through something that returned an int, for instance.

Reply:
Yeah...and we only documented that the function returns real. I am fine with that restriction.

Reply:
Cool cool. Will resolve this then.